File: Problem90_SubsetsII.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<vector<int>> res;
    vector<int> oneSub;
    void backtrack(int start, vector<int>& nums) {
        res.push_back(oneSub);
        for (int i=start;i<nums.size();i++) {
            if (i > start && nums[i]==nums[i-1]) continue;
            oneSub.push_back(nums[i]);
            backtrack(i+1, nums);
            oneSub.pop_back();
        }
    }
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        backtrack(0, nums);
        return res;
    }
};

int main() {
    Solution solve;
    vector<int> nums = {4,4,4,1,4};
    vector<vector<int>> res = solve.subsetsWithDup(nums);
    for (auto& row : res) {
        for (int x : row) cout << x << " ";
        cout << endl;
    }
}

File: Problem90_SubsetsII.py
from typing import List

class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        res, subRes = [], []
        def backtrack(start):
            res.append(subRes.copy())
            for i in range(start, len(nums)):
                if i>start and nums[i]==nums[i-1]: continue
                subRes.append(nums[i])
                backtrack(i+1)
                subRes.pop()
        nums.sort()
        backtrack(0)
        return res

if __name__=="__main__":
    solve = Solution()
    print(solve.subsetsWithDup([1,2,2]))
    print(pow(2,16))